rfqservice/negotiator.go

```diff
@@ -0,0 +1,279 @@
+package rfqservice
+
+import (
+	"encoding/hex"
+	"fmt"
+	"sync"
+
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/lightninglabs/taproot-assets/asset"
+	"github.com/lightninglabs/taproot-assets/fn"
+	"github.com/lightninglabs/taproot-assets/rfqmsg"
+)
+
+// NegotiatorCfg holds the configuration for the negotiator.
+type NegotiatorCfg struct {
+	// PriceOracle is the price oracle that the negotiator will use to
+	// determine whether a quote is accepted or rejected.
+	PriceOracle PriceOracle
+}
+
+// Negotiator is a struct that handles the negotiation of quotes. It is a RFQ
+// subsystem. It determines whether a quote is accepted or rejected.
+type Negotiator struct {
+	startOnce sync.Once
+	stopOnce  sync.Once
+
+	// cfg holds the configuration parameters for the negotiator.
+	cfg NegotiatorCfg
+
+	// outgoingMessages is a channel which is populated with outgoing peer
+	// messages.
+	outgoingMessages chan<- rfqmsg.OutgoingMsg
+
+	// assetSellOffers is a map (keyed on asset) that holds the asset sell
+	// offers that the negotiator has received.
+	assetSellOffers map[string]AssetSellOffer
+
+	assetGroupSellOffers map[string]AssetSellOffer
+
+	// ContextGuard provides a wait group and main quit channel that can be
+	// used to create guarded contexts.
+	*fn.ContextGuard
+}
+
+// NewNegotiator creates a new quote negotiator.
+func NewNegotiator(cfg NegotiatorCfg,
+	outgoingMessages chan<- rfqmsg.OutgoingMsg) (*Negotiator, error) {
+
+	return &Negotiator{
+		cfg:              cfg,
+		outgoingMessages: outgoingMessages,
+		ContextGuard: &fn.ContextGuard{
+			DefaultTimeout: DefaultTimeout,
+			Quit:           make(chan struct{}),
+		},
+	}, nil
+}
+
+// queryPriceOracle queries the price oracle for the asking price.
+func (n *Negotiator) queryPriceOracle(
+	req rfqmsg.Request) (*OracleAskingPriceResponse, error) {
+
+	ctx, cancel := n.WithCtxQuitNoTimeout()
+	defer cancel()
+
+	res, err := n.cfg.PriceOracle.QueryAskingPrice(
+		ctx, req.AssetID, req.AssetGroupKey, req.AssetAmount,
+		&req.BidPrice,
```

Comment:
Why's a bit price needed? In both directions, the receiver/sender is the one that's asking the edge node for a price quote. Or is this the hint price that it sends along?

Reply:
This is the request bid price. It's the hint price provided by the peer which is requesting a quote. I thought the price oracle might as well have access to it. But we might not want to allow that because the quote requesting peer might be colluding with the price oracle.